### PR TITLE
Conditions for order and step change in NDFs

### PR DESCRIPTION
**Pull request by utkarsh530**
Hi,

I have made some changes referring to the paper [here](https://www.sciencedirect.com/science/article/pii/S1877050914002683). It states that order and step-size must not be changed for at least order+2 iterations, which was not implemented I guess. Please have a look. Also, Shampine's Paper only mentions about error estimator, it does not say anything about order and step-size changes, hence referring the paper. 
@ChrisRackauckas @kanav99

### CODE REVIEW

**Comment by ChrisRackauckas** [comment]
How are the benchmarks looking?

**Comment by utkarsh530** [comment]
```

abstols = 1.0 ./ 10.0 .^ (5:8)
reltols = 1.0 ./ 10.0 .^ (1:4);
setups = [Dict(:alg=>Rosenbrock23()),
          Dict(:alg=>Rodas3()),
          Dict(:alg=>TRBDF2()),
          Dict(:alg=>Rodas5()),
          Dict(:alg=>RadauIIA5()),
          Dict(:alg=>lsoda()),
          Dict(:alg=>RadauIIA5()),
          Dict(:alg=>QNDF()),
          ]
wp = WorkPrecisionSet(prob,abstols,reltols,setups;verbose=false,
                      save_everystep=false,appxsol=test_sol,maxiters=Int(1e5),numruns=10)
```
wp_updated

**Comment by ChrisRackauckas** [comment]
before and after?

**Comment by utkarsh530** [comment]
@ChrisRackauckas I updated the comment.
One thing would like to point out:
```
julia> function test!(du,u,p,t)
            du[1] = -10*u[1]
            du[2] = -0.01*u[2]
            end
test! (generic function with 1 method)

julia> 

julia> u0 = [1.01;1.01]
2-element Array{Float64,1}:
 1.01
 1.01

julia> tspan = (0.0, 1.0)
(0.0, 1.0)

julia> sol = solve(prob, QNDF(),dt=0.003577708769278081,reltol=1/10^3,abstol=1/10^6)
ERROR: UndefVarError: prob not defined
Stacktrace:
 [1] top-level scope at REPL[6]:1
 [2] eval(::Module, ::Any) at .\boot.jl:330
 [3] eval_user_input(::Any, ::REPL.REPLBackend) at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.3\REPL\src\REPL.jl:86
 [4] run_backend(::REPL.REPLBackend) at C:\Users\Utkarsh\.julia\packages\Revise\Pcs5V\src\Revise.jl:1073
 [5] top-level scope at none:0

julia> prob = ODEProblem(test!, u0, tspan)
ODEProblem with uType Array{Float64,1} and tType Float64. In-place: true
timespan: (0.0, 1.0)
u0: [1.01, 1.01]

julia> sol = solve(prob, QNDF(),dt=0.003577708769278081,reltol=1/10^3,abstol=1/10^6)
t = 0.0
cnt = 1
integrator.cache.h = 0.0
t = 0.003577708769278081
cnt = 2
integrator.cache.h = 0.0
t = 0.007155417538556162
cnt = 3
integrator.cache.h = 0.0057065046428849 #h is computed this there
t = 0.010733126307834243 #why t is changing with dt=0.003577708769278081?
cnt = 4
integrator.cache.h = 0.0057065046428849
t = 0.016439630950719146
cnt = 5
integrator.cache.h = 0.0057065046428849
t = 0.022146135593604045
cnt = 6
integrator.cache.h = 0.0057065046428849
t = 0.027852640236488944
cnt = 7
integrator.cache.h = 0.0213581753838296
t = 0.033559144879373844
cnt = 8
integrator.cache.h = 0.0213581753838296
t = 0.05491732026320344
cnt = 9
integrator.cache.h = 0.0213581753838296
t = 0.07627549564703304
cnt = 10
integrator.cache.h = 0.0213581753838296
t = 0.09763367103086264
cnt = 11
integrator.cache.h = 0.0213581753838296
t = 0.11899184641469224
cnt = 12
t = 0.11899184641469224
cnt = 13
integrator.cache.h = 0.015753246404189358
t = 0.12967093410660704
cnt = 14
integrator.cache.h = 0.015753246404189358
t = 0.1454241805107964
cnt = 15
integrator.cache.h = 0.015753246404189358
t = 0.16117742691498574
cnt = 16
integrator.cache.h = 0.015753246404189358
t = 0.17693067331917509
cnt = 17
integrator.cache.h = 0.03287987110603007
t = 0.19268391972336443
cnt = 18
integrator.cache.h = 0.03287987110603007
t = 0.2255637908293945
cnt = 19
integrator.cache.h = 0.03287987110603007
t = 0.25844366193542456
cnt = 20
integrator.cache.h = 0.03287987110603007
t = 0.29132353304145464
cnt = 21
integrator.cache.h = 0.03287987110603007
t = 0.3242034041474847
cnt = 22
integrator.cache.h = 0.03287987110603007
t = 0.3570832752535148
cnt = 23
t = 0.3570832752535148
cnt = 24
integrator.cache.h = 0.023728827392534308
t = 0.37352321080652984
cnt = 25
integrator.cache.h = 0.023728827392534308
t = 0.39725203819906413
cnt = 26
integrator.cache.h = 0.023728827392534308
t = 0.4209808655915984
cnt = 27
integrator.cache.h = 0.023728827392534308
t = 0.4447096929841327
cnt = 28
integrator.cache.h = 0.023728827392534308
t = 0.468438520376667
cnt = 29
integrator.cache.h = 0.034727869668566155
t = 0.4921673477692013
cnt = 30
integrator.cache.h = 0.034727869668566155
t = 0.5268952174377675
cnt = 31
integrator.cache.h = 0.034727869668566155
t = 0.5616230871063336
cnt = 32
integrator.cache.h = 0.034727869668566155
t = 0.5963509567748997
cnt = 33
integrator.cache.h = 0.034727869668566155
t = 0.6310788264434658
cnt = 34
integrator.cache.h = 0.034727869668566155
t = 0.6658066961120319
cnt = 35
t = 0.6658066961120319
cnt = 36
integrator.cache.h = 0.028808977460981638
t = 0.683170630946315
cnt = 37
integrator.cache.h = 0.028808977460981638
t = 0.7119796084072967
cnt = 38
integrator.cache.h = 0.028808977460981638
t = 0.7407885858682783
cnt = 39
integrator.cache.h = 0.028808977460981638
t = 0.76959756332926
cnt = 40
integrator.cache.h = 0.028808977460981638
t = 0.7984065407902416
cnt = 41
integrator.cache.h = 0.042835951471624195
t = 0.8272155182512233
cnt = 42
integrator.cache.h = 0.042835951471624195
t = 0.8700514697228475
cnt = 43
integrator.cache.h = 0.042835951471624195
t = 0.9128874211944717
cnt = 44
integrator.cache.h = 0.042835951471624195
t = 0.955723372666096
cnt = 45
integrator.cache.h = 0.042835951471624195
t = 0.9985593241377202
cnt = 46
integrator.cache.h = 0.042835951471624195
retcode: Success
Interpolation: 3rd order Hermite
t: 44-element Array{Float64,1}:
 0.0
 0.003577708769278081
 0.007155417538556162
 0.010733126307834243
 ⋮
 0.9128874211944717
 0.955723372666096
 0.9985593241377202  
 1.0
u: 44-element Array{Array{Float64,1},1}:
 [1.01, 1.01]
 [0.9751132864406183, 1.0099638664341841]
 [0.9412428493295016, 1.0099277339520365]
 [0.9085271261969265, 1.0098916027395566]
 ⋮
 [0.00011102720133726845, 1.0008217908885546]
 [7.26730378226713e-5, 1.0003931693191954]
 [4.7705764144002284e-5, 0.9999647336271009]
 [4.701888852288306e-5, 0.9999503275032385]
```
Please see the comments I made in the REPL snippet? Why is this happening?

**Comment by ChrisRackauckas** [comment]
>Please see the comments I made in the REPL snippet after solving QNDF and logging the details. Why is this happening?

What comment?

>I am trying to move this code in completely in controllers.jl. How to write step_accept and step_reject controllers? How do they work?

The you essentially just write something that changes dt. The best example is the "standard controller"

**Comment by ChrisRackauckas** [comment]
>Could you point out where other than this code is using integrator.EEst to estimate dt?

That is in the controller section.

**Comment by kanav99** [comment]
> Please see the comments I made in the REPL snippet after solving QNDF and logging the details. Why is this happening?

Probably faulty controller. Last time I saw QNDF implementation, the `h` and `dt` seemed to be out of sync at many places. Best thing would be to make a new controller instead of tweaking with the existing implementation. That would be much easier and more maintainable.

> I am trying to move this code in completely in controllers.jl

Awesome!

>  Could you point out where other than this code is using integrator.EEst to estimate dt?

you get the `q` from controller here https://github.com/SciML/OrdinaryDiffEq.jl/blob/master/src/integrators/integrator_utils.jl#L185
which is then used to calculate `dt` (that is just some lines below that line)

**Comment by ChrisRackauckas** [comment]
Rebase.

File: src/integrators/controllers.jl [hunk 1]
```diff
@@ -138,11 +138,21 @@ function stepsize_controller!(integrator, alg::QNDF)
   cnt = integrator.iter
   EEst1 = integrator.cache.EEst1
   EEst2 = integrator.cache.EEst2
-  if cnt <= 3
+  if integrator.cache.nconsteps < integrator.cache.order + 2
     # q = standard_stepsize_controller!(integrator, alg)
+    #@show("Here")
+    #@show integrator.cache.nconsteps
```
Comment (ChrisRackauckas) [inline src/integrators/controllers.jl:144]: cleanup

File: src/integrators/controllers.jl [hunk 2]
```diff
@@ -138,11 +138,21 @@ function stepsize_controller!(integrator, alg::QNDF)
...
+    #@show integrator.cache.order
+    integrator.cache.nconsteps += 1
```
Comment (ChrisRackauckas) [inline src/integrators/controllers.jl:146]:
```suggestion
    integrator.cache.nconsteps += 1
```

**Comment by ChrisRackauckas** [comment]
Possibly one thing that violates `nconsteps` is if you get near a tstop, the `dt` will drop so you don't step over the end of the interval. So a dt change is possible even when the algorithm thinks it's not possible because it doesn't change dt. It does look like the implementation is robust to this though.

**Comment by utkarsh530** [comment]
> Possibly one thing that violates `nconsteps` is if you get near a tstop, the `dt` will drop so you don't step over the end of the interval. So a dt change is possible even when the algorithm thinks it's not possible because it doesn't change dt. It does look like the implementation is robust to this though.

Thank you for your insights @YingboMa @ChrisRackauckas !
Any pointers on how to fix this? Like previously before NDF controller was merged, this was not the problem then. Why is it happening ow?

**Comment by ChrisRackauckas** [comment]
Are you sure it wasn't always a problem?

**Comment by utkarsh530** [comment]
I only tested with that problem only, it was fine back then as posted [here](https://github.com/SciML/OrdinaryDiffEq.jl/pull/1130#issuecomment-625981359) not sure completely.

**Comment by ChrisRackauckas** [comment]
Can you show how you're identifying the problem? I don't directly see it?

**Comment by utkarsh530** [comment]
It's just on observational basis as per se, it's not a problem although because it's happening according to the code. But what is happening is that it wrongly calculates for eg. `t=1.0` (which is an observation) in this current PR, which was not happening earlier particular in that test problem, that baffles me. Like you mentioned that while my controller is trying to keep the time step constant, the integrator will shorten `dt`. I agree that it is robust though, which keeps steps-quasi constant, just wanted to know how to handle that shortcoming I mentioned earlier.

**Comment by ChrisRackauckas** [comment]
> But what is happening is that it wrongly calculates for eg. t=1.0

What do you mean? Can you show me how you're seeing this?

**Comment by utkarsh530** [comment]
I don't mean wrongly, just an observation in this problem:

```
function test!(du, u, p, t)
    du[1] = -10 * u[1]
    du[2] = -0.01 * u[2]
end

u0 = [1.01;1.01]
tspan = (0.0, 1.0)
prob = ODEProblem(test!, u0, tspan) 
```
Analytically, the solution at `t=1.0` should be `[0.000045839, 0.999950332]`. The previous build before merging NDF controller was showing `[4.701888852288306e-5, 0.9999503275032385]`, which was quite close to that, now as I posted earlier is `[0.00010127652098422391, 0.999951251834322]`. Just wanted to know why is this happening? It's just an observation, it's working correctly as it must be, but why this shortfall, what changes must be done if there are some?

**Comment by ChrisRackauckas** [comment]
The default solver tolerance is relative 1e-3, so it's only trying to get the first 3 digits correct. Absolute tolerance is 1e-6, which is trying to get 6 digits when small. Those changes are below tolerance so I wouldn't worry about them (as long as it converges as you decrease the tolerance!)

**Comment by utkarsh530** [comment]
I completely agree with that, and that makes sense. Thank you for your patience and trying to clear my doubts. 😄

**Comment by ChrisRackauckas** [comment]
Awesome. How are the BDF methods doing on the new master?

**Comment by utkarsh530** [comment]
qndf_latest

**Comment by ChrisRackauckas** [comment]
that's looking a ton better though. But hmm... still not the best. When you profile, where do you see the time being spent?

**Comment by utkarsh530** [comment]
I did not get that. You mean where the most time is spent in `perform_step!` by marking every line with `@time` or are you talking about allocations?

**Comment by ChrisRackauckas** [comment]
See https://stackoverflow.com/questions/49719076/macos-python-with-numpy-faster-than-julia-in-training-neural-network/49724611#49724611 for information on profiling.

**Pull request merged**